### PR TITLE
Database view emulator. ActiveRecord for query statement.

### PR DESCRIPTION
**Pull request by mdmunir**
| Q | A |
| --- | --- |
| Is bugfix? | no |
| New feature? | yes |
| Breaks BC? | no |
| Tests pass? | yes |
| Fixed issues | #12025 |

### CODE REVIEW

File: framework/db/QueryRecord.php [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+ *     echo $model->paid->total;
+ * }
+ * ```
+ * For more details and usage information on ActiveRecord, see the [guide article on ActiveRecord](guide:db-active-record).
+ *
+ * @method ActiveQuery hasMany($class, array $link) see [[BaseActiveRecord::hasMany()]] for more info
+ * @method ActiveQuery hasOne($class, array $link) see [[BaseActiveRecord::hasOne()]] for more info
+ *
+ * @author Misbahul D Munir <misbahuldmunir@gmail.com>
+ * @since 2.10
```
Comment (samdark) [inline framework/db/QueryRecord.php]: 2.0.10

File: framework/db/QueryRecord.php [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+ */
+class QueryRecord extends BaseActiveRecord
```
Comment (samdark) [inline framework/db/QueryRecord.php]: Should it be abstract?

File: framework/db/QueryRecord.php [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+     */
+    public static function getDb()
+    {
+        return Yii::$app->getDb();
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public static function primaryKey()
```
Comment (samdark) [inline framework/db/QueryRecord.php]: Could be left w/o implementation.

File: tests/framework/db/QueryRecordTest.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+use yiiunit\data\ar\ActiveRecord;
+use yiiunit\data\ar\QueryRecord;
+use yiiunit\data\ar\TotalOrder;
+use yiiunit\data\ar\Item;
+
+/**
+ * Description of QueryRecordTest
+ *
+ * @author Misbahul D Munir <misbahuldmunir@gmail.com>
+ * @since 1.0
```
Comment (samdark) [inline tests/framework/db/QueryRecordTest.php]: ?

File: framework/db/QueryRecord.php (hunk 2 above)
Comment (mdmunir) [inline framework/db/QueryRecord.php]: ya, agree.
but when create unit test, raise error `should not be abstract`.

File: tests/framework/db/QueryRecordTest.php (hunk 4 above)
Comment (mdmunir) [inline tests/framework/db/QueryRecordTest.php]: autogonerate from netbeans :D.

File: framework/db/QueryRecord.php (hunk 3 above)
Comment (mdmunir) [inline framework/db/QueryRecord.php]: > PHP Fatal error:  Class yiiunit\data\ar\TotalOrder contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (yii\db\ActiveRecordInterface::primaryKey) in /home/mdmunir/projects/php/mdmunir/yii2/tests/data/ar/TotalOrder.php on line 25

File: framework/db/QueryRecord.php (hunk 2 above)
Comment (mdmunir) [inline framework/db/QueryRecord.php]: disagree.
This class should can be implemented with minimal code. I mean, just implement `query()` and `attributes()`.

File: framework/db/QueryRecord.php (hunk 3 above)
Comment (samdark) [inline framework/db/QueryRecord.php]: Well, yes. `TotalOrder` needs to implement all required methods.

**Comment by samdark** [comment]
@klimov-paul what do you think about it?

**Comment by klimov-paul** [comment]
> @klimov-paul what do you think about it?

This should not belong to the core.

As I already mentioned at th issue, there is no necessity of such thing as it can be easily achieved via default scopes feature:
https://github.com/yiisoft/yii2/issues/12025#issuecomment-235575887

Of course, there could be a situation, when such trick will not work, however I consider this to be a rare case.

I suggest this solution should be composed into an extension and maintained out of 'yiisoft' scope.

**Comment by cebe** [comment]
I agree with @klimov-paul. @mdmunir feel free to publish it as an extension.

**Comment by mdmunir** [comment]
As i said, this is View Emulator. I mean, everithing that can be do using view, can be done using this class. Some people prefer use `view` another people prefer using sql directly. If i have ability to create view in aplication, i chose create it in aplication.

@klimov-paul code not work for situation like this
```php
Invoice::find()
    ->alias('i')
    ->joinWith(['paid p'])
    ->where('[[i.value]] > [[p.total]]') // must use having instead of where
```
^ that code not a rare case (for me). That the reason why i need this class.

Simple way to create the same class without `ViewRecord` is directly inherit from `BaseActiveRecord`. But, if that so, i must override methods `getDb()`, `insert()`, `primaryKey()` many many times because that methods are abstract. 

> feel free to publish it as an extension.

I guess my own extension. Ya, i use this in my project (not as extension, but as a single class in my project). I submit it to core yii just to share my code to other yii user.

**Pull request closed without merging**